subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java

```diff
@@ -40,39 +41,47 @@
     private static final Logger LOGGER = Logging.getLogger(SkipEmptySourceFilesTaskExecuter.class);
     private final TaskInputsListener taskInputsListener;
     private final BuildOutputCleanupRegistry buildOutputCleanupRegistry;
+    private final TaskOutputsGenerationListener taskOutputsGenerationListener;
     private final TaskExecuter executer;
 
-    public SkipEmptySourceFilesTaskExecuter(TaskInputsListener taskInputsListener, BuildOutputCleanupRegistry buildOutputCleanupRegistry, TaskExecuter executer) {
+    public SkipEmptySourceFilesTaskExecuter(TaskInputsListener taskInputsListener, BuildOutputCleanupRegistry buildOutputCleanupRegistry, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskExecuter executer) {
         this.taskInputsListener = taskInputsListener;
         this.buildOutputCleanupRegistry = buildOutputCleanupRegistry;
+        this.taskOutputsGenerationListener = taskOutputsGenerationListener;
         this.executer = executer;
     }
 
     public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
         FileCollection sourceFiles = task.getInputs().getSourceFiles();
         if (task.getInputs().getHasSourceFiles() && sourceFiles.isEmpty()) {
             TaskArtifactState taskArtifactState = context.getTaskArtifactState();
-            FileCollection outputFiles = taskArtifactState.getExecutionHistory().getOutputFiles();
+            TaskExecutionHistory executionHistory = taskArtifactState.getExecutionHistory();
+            Set<File> outputFiles = executionHistory.getPreviousOutputs();
             if (outputFiles == null) {
                 state.setOutcome(TaskExecutionOutcome.NO_SOURCE);
                 LOGGER.info("Skipping {} as it has no source files and no history of previous output files.", task);
             } else if (outputFiles.isEmpty()) {
                 state.setOutcome(TaskExecutionOutcome.NO_SOURCE);
                 LOGGER.info("Skipping {} as it has no source files and no previous output files.", task);
             } else {
-                Set<File> outputFileSet = outputFiles.getFiles();
+                boolean cleanupDirectories = true;
+                if (executionHistory.getOverlappingOutputs() != null) {
+                    LOGGER.info("No leftover directories for {} will be deleted since overlapping outputs where detected.", task);
+                    cleanupDirectories = false;
+                }
+                taskOutputsGenerationListener.beforeTaskOutputsGenerated();
```

Comment:
Actually, it should make things faster since on the second run we don't have any output files to clean up. The cleanup may be a little bit faster but will happen only once.